pkg/volume/plugins.go

```diff
@@ -216,6 +216,32 @@ type ExpandableVolumePlugin interface {
 	RequiresFSResize() bool
 }
 
+// VolumePluginWithAttachLimits is an extended interface of VolumePlugin that restricts number of
+// volumes that can be attached to a node.
+type VolumePluginWithAttachLimits interface {
+	VolumePlugin
+	// Return maximum number of volumes that can be attached to a node for this plugin.
+	// The key must be same as string returned by VolumeLimitKey function. The returned
+	// map may look like:
+	//     - { "storage-limits-aws-ebs": 39 }
+	//     - { "storage-limits-gce-pd": 10 }
+	// A volume plugin may return error from this function - if it can not be used on a given node or not
+	// applicable in given environment (where environment could be cloudprovider or any other dependency)
+	// For example - calling this function for EBS volume plugin on a GCE node should
+	// result in error.
+	// The returned values are stored in node allocatable property and will be used
+	// by scheduler to determine how many pods with volumes can be scheduled on given node.
+	GetVolumeLimits() (map[string]int64, error)
+	// Return volume limit key string to be used in node capacity constraints
+	// The key must start with prefix storage-limits-. For example:
+	//    - storage-limits-aws-ebs
+	//    - storage-limits-csi-cinder
+	// The key should respect character limit of ResourceName type
+	// This function may be called by kubelet or scheduler to identify node allocatable property
```

Comment:
how does the scheduler call volume plugin methods?

Reply:
Addressed below in response to a comment by @bsalamat  https://github.com/kubernetes/kubernetes/pull/64154#discussion_r190767908

Reply:
fixed. removed volume plugin manager from scheduler. We are probably never going to support dynamic volume limits for flex. We will revisit the problem for CSI in 1.12.